### PR TITLE
Fix README and sstate issues

### PR DESCRIPTION
**Pull request by liuming50**

### CODE REVIEW

File: classes/image_type_tezi_run.bbclass [hunk 1]
```diff
@@ -1,26 +1,31 @@
-inherit image_types
+python () {
+    if bb.utils.contains("IMAGE_FSTYPES", "tezirunimg", "1", "0", d) == "1":
+        bb.build.addtask('do_assemble_tezirunimg', 'do_image_complete', 'do_image_squashfs', d)
+}
```
Comment (agners) [inline classes/image_type_tezi_run.bbclass]: This adds a new task which is before `do_image_complete` and after `do_image_squashfs`, right?

Isn't that what a image type would do with squashfs as dependency (e.g. what we had before with `IMAGE_TYPEDEP_tezirunimg += "squashfs"`)? I don't quite get the difference here, it seems to me a bit of misuse of the image_type when we don't use `IMAGE_CMD_...` and friends...

File: classes/image_type_tezi_run.bbclass [hunk 2]
```diff
@@ -1,26 +1,31 @@
...
 
-do_image_tezirunimg[depends] += "tezi-run-metadata:do_deploy u-boot-mkimage-native:do_install p7zip-native:do_install"
 TEZI_DISTRO_BOOT_SCRIPTS ??= "boot-sdp.scr boot.scr"
 UBOOT_BINARY ??= "u-boot.${UBOOT_SUFFIX}"
 TEZI_UBOOT_BINARY_EMMC ??= "${UBOOT_BINARY}"
 TEZI_UBOOT_BINARY_RAWNAND ??= "${UBOOT_BINARY}"
 TEZI_UBOOT_BINARY_RECOVERY ??= "${UBOOT_BINARY}"
+TEZI_UBOOT_BINARIES ??= "${@' '.join(x for x in sorted(set([bb.utils.contains('TORADEX_FLASH_TYPE', 'emmc', d.getVar('TEZI_UBOOT_BINARY_EMMC', True), '', d), \
+                         bb.utils.contains('TORADEX_FLASH_TYPE', 'rawnand', d.getVar('TEZI_UBOOT_BINARY_RAWNAND', True), '', d), \
+                         d.getVar('TEZI_UBOOT_BINARY_RECOVERY', True)])))}"
```
Comment (agners) [inline classes/image_type_tezi_run.bbclass]: Populating `TEZI_UBOOT_BINARIES` outside of python is nicer, can you split this change out into a separate commit? Than we can merge that on its own.

File: classes/image_type_tezi_run.bbclass [hunk 3]
```diff
@@ -150,74 +154,53 @@ def rootfs_tezi_run_create_json(d, flash_type, type_specific_name = False):
...
+	# Copy device tree
+	for dtb in ${KERNEL_DEVICETREE}; do
+		cp ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE}-$dtb ${WORKDIR}/tmp-fitimage
+	done
+
+	# Copy rootfs image
+	if [ -e ${IMGDEPLOYDIR}/${IMAGE_BASENAME}-${MACHINE}.squashfs ]; then
+		cp ${IMGDEPLOYDIR}/${IMAGE_BASENAME}-${MACHINE}.squashfs ${WORKDIR}/tmp-fitimage
+	else
+		cp ${DEPLOY_DIR_IMAGE}/${IMAGE_BASENAME}-${MACHINE}.squashfs ${WORKDIR}/tmp-fitimage
```
Comment (agners) [inline classes/image_type_tezi_run.bbclass]: That looks a bit hacky :-)

File: classes/image_type_tezi_run.bbclass [hunk 4]
```diff
@@ -150,74 +154,53 @@ def rootfs_tezi_run_create_json(d, flash_type, type_specific_name = False):
...
-
-addtask do_assemble_fitimage after do_image_complete before do_build
-do_assemble_fitimage[depends] = "tezi-run-metadata:do_deploy"
-
-IMAGE_TYPEDEP_tezirunimg += "squashfs"
+do_assemble_tezirunimg[dirs] = "${WORKDIR}/tmp-fitimage ${WORKDIR}/tmp-tezirunimage"
+do_assemble_tezirunimg[cleandirs] = "${WORKDIR}/tmp-fitimage ${WORKDIR}/tmp-tezirunimage"
+do_assemble_tezirunimg[depends] = "virtual/bootloader:do_deploy u-boot-distro-boot:do_deploy virtual/kernel:do_deploy \
+                                 tezi-run-metadata:do_deploy u-boot-mkimage-native:do_populate_sysroot \
+                                 p7zip-native:do_populate_sysroot zip-native:do_populate_sysroot \
```
Comment (agners) [inline classes/image_type_tezi_run.bbclass]: I guess those dependency fixes are somewhat independent on the image type change. Can you also create a separate patch which fixes `do_image_tezirunimg[depends]`?

**Review by agners — CHANGES_REQUESTED** [review]
Hi Ming,

Thanks for your improvements! Some review though. Especially, can you split out the non-controversial part so we can get this merged soon?

Thanks!

File: classes/image_type_tezi_run.bbclass (hunk 1 above)
Comment (liuming50) [reply classes/image_type_tezi_run.bbclass]: Yes, usually it could be handled by IMAGE_TYPEDEP, but the tricky part here is, the do_image_tezirunimg task is a bundle of python tasks and shell tasks, but IMAGE_CMD only support shell tasks, so we can not make do_image_tezirunimg a IMAGE_CMD, and we also can not make it to be a empty IMAGE_CMD as well, since there are quite some processes in image.bbclass to add prefuncs and so on to IMAGE_CMD, they could not work with RM_WORK, for instance, the set_image_size task requires IMAGE_ROOTFS, but it does not always exist since RM_WORK is being enabled, which means the IMAGE_CMD might fail if its task hash changed(or its dependent tasks' hashes) while the do_rootfs task hash not changed. So we have to mask the processes in image.bbclass with IMAGE_TYPES_MASKED += "tezirunimg", so that's why we need add the task dependency do_assemble_tezirunimg -> do_image_squashfs in this way.

File: classes/image_type_tezi_run.bbclass (hunk 2 above)
Comment (liuming50) [reply classes/image_type_tezi_run.bbclass]: Yes, sure, will split it in V2.

File: classes/image_type_tezi_run.bbclass (hunk 3 above)
Comment (liuming50) [reply classes/image_type_tezi_run.bbclass]: Yes, it is. I can explain the idea here.

In the original implementation, why the artifact could not be covered by sstate is we were adding do_assemble_tezirunimg after do_image_complete, and why we were doing that is because do_assemble_tezirunimg requires kernel images, dtbs, and squashfs in the same directory with its file, to assemble the fitimage, so, to be able to make all the files in the same directories with its file, we copy them into ${WORKDIR}/tmp-fitimage, including the squashfs image, and that file could be untared from sstate (in DEPLOY_DIR_IMAGE) or generated by do_image_squashfs (in IMGDEPLOYDIR).

Comment (liuming50) [reply classes/image_type_tezi_run.bbclass]: Yes, I agree, this part is really tricky.

File: classes/image_type_tezi_run.bbclass (hunk 4 above)
Comment (liuming50) [reply classes/image_type_tezi_run.bbclass]: Sure, will do that in V2.

File: classes/image_type_tezi_run.bbclass (hunk 3 above)
Comment (agners) [reply classes/image_type_tezi_run.bbclass]: I see what you are saying, but when we have a `IMAGE_TYPEDEP_tezirunimg` it is somehow guaranteed that the squashfs is in the right place...

Can we maybe somehow enable python with IMAGE_CMD?

**Comment by liuming50** [comment]
OK, will think about that.

//Ming Liu

Stefan Agner <notifications@github.com> 於 2018年12月7日 週五 下午4:19寫道：

**Comment by liuming50** [comment]
Hi, Stefan:

I have split the commits and modified to use the IMAGE_TYPEDEP_tezirunimg as you suggested, please kindly help me review them.

**Comment by agners** [comment]
Thanks for the update! Patch 1-3 look good, I will apply them.

However, with the last one I am still not quite happy. Two questions:
Why do you think we need the change? You say that it is sstate tracked, but do we care? What is the advantage?

It now feels a bit hacky that we have to omit rm_work... Can we maybe do it entirly differently by creating the fit image as part of the kernel recipe, similar to how it is done in `openembedded-core/meta/classes/kernel-fitimage.bbclass`? This probably would need that the kernel depends on the image, which is a bit wired, but could work?

**Comment by liuming50** [comment]
Hi, Stefan:

I am on board the changes are really hacky, since we are doing a fitimage
with the normal rootfs, and the paths of kernel/dtb/rootfs are being
hard-coded in .its file, hence the assembling task would reply on the
output of its dependent tasks to be present when it executes...and so on

I am totally on board, the better way to do this is inheriting
kernel-fitimage.bbclass

//Ming Liu



Stefan Agner <notifications@github.com> 於 2018年12月11日 週二 上午9:41寫道：
